File: frontend/src/types/historian.ts
export type TagName = string;
export type RetrievalMode = "raw" | "delta" | "cyclic";
export type OutputFormat = "csv" | "xlsx";
export type ScalarValue = string | number | boolean | null;
export type PreviewColumn = string;
export type TagSystem =
  | "Gen 2"
  | "DeltaV BMS/FARC"
  | "SFR"
  | "Historian Internal"
  | "Unknown";

export interface TagMetadata {
  tag_name?: TagName | null;
  name?: TagName | null;
  description?: string | null;
  io_address?: string | null;
  units?: string | null;
  engineering_unit?: string | null;
  source_system?: string | null;
  system?: TagSystem | null;
  is_system_tag?: boolean | null;
}

export interface HistorianQuery {
  start_datetime: string;
  end_datetime: string;
  tags: TagName[];
  retrieval_mode: RetrievalMode;
  resolution_milliseconds?: number | null;
}

export type PreviewRequest = HistorianQuery;

export interface ExportRequest extends HistorianQuery {
  output_format: OutputFormat;
}

export interface PreviewRow {
  timestamp: string;
  values: Record<TagName, ScalarValue>;
}

export interface PreviewResponse {
  columns: PreviewColumn[];
  rows: PreviewRow[];
}

File: backend/app/domain/historian.py
from enum import StrEnum

TIMESTAMP_COLUMN = "timestamp"
DEFAULT_EXPORT_BASENAME = "historian-export"
MAX_PREVIEW_ROWS = 200


class RetrievalMode(StrEnum):
    raw = "raw"
    delta = "delta"
    cyclic = "cyclic"


class OutputFormat(StrEnum):
    csv = "csv"
    xlsx = "xlsx"


TagName = str
ScalarValue = float | int | str | bool | None

EXPORT_FILENAMES: dict[OutputFormat, str] = {
    OutputFormat.csv: f"{DEFAULT_EXPORT_BASENAME}.csv",
    OutputFormat.xlsx: f"{DEFAULT_EXPORT_BASENAME}.xlsx",
}


def derive_source_system(io_address: str | None) -> str | None:
    if not io_address:
        return None

    normalized_address = io_address.strip()
    if not normalized_address:
        return None

    for delimiter in ("::", ".", ":", "/"):
        if delimiter in normalized_address:
            source_system = normalized_address.split(delimiter, 1)[0].strip()
            return source_system or None

    return normalized_address
